src/G2/Equiv/Tactics.hs

```diff
@@ -0,0 +1,785 @@
+        Just (hm, _) -> hm
+      maybe_pairs = map mr prev
+      obs_sets = map getObs maybe_pairs
+      h1 = expr_env s1
+      h2 = expr_env s2
+      obs_sets' = map (HS.filter (\(e1, e2) -> rfs h1 e1 && rfs h2 e2)) obs_sets
+      no_loss = map (\(hs1, hs2) -> HS.size hs1 == HS.size hs2) (zip obs_sets obs_sets')
+      mpc m = case m of
+        (Just (hm, _), (s_old1, s_old2, _)) ->
+          andM (moreRestrictivePC solver s_old1 s1 hm) (moreRestrictivePC solver s_old2 s2 hm)
+        _ -> return False
+      bools = map mpc (zip maybe_pairs prev)
+  -- check obligations individually rather than as one big group
+  res_list <- W.liftIO $ mapM (checkObligations solver s1 s2) obs_sets'
+  bools' <- mapM id bools
+  {-
+  print "#####"
+  print $ folder_name $ track s1
+  print $ folder_name $ track s2
+  print "PREV"
+  print $ map (\(p1, p2, _) -> (folder_name $ track p1, folder_name $ track p2)) prev
+  print "MPAIRS"
+  print $ map isJust maybe_pairs
+  print "BOOLS"
+  print bools'
+  -}
+  -- need res_list, no_loss, and bools all aligning at a point
+  let all_three thr = case fst thr of
+        ((S.UNSAT (), _), (True, True)) -> True
+        _ -> False
+  -- all four lists should be the same length
+  case filter all_three $ zip (zip (zip res_list prev) $ zip no_loss bools') maybe_pairs of
+    [] -> return Nothing
+    (((_, (p1, p2, pc)), _), m):_ -> return $ Just $ PrevMatch (s1, s2) (p1, p2) (getMap m, getObs m) pc
+
+-- the third entry in prev tuples is meaningless here
+moreRestrictivePair :: S.Solver solver =>
+                       solver ->
+                       HS.HashSet Name ->
+                       [(StateET, StateET)] ->
+                       (StateET, StateET) ->
+                       W.WriterT [Marker] IO (Maybe (PrevMatch EquivTracker))
+moreRestrictivePair solver ns prev (s1, s2) =
+  let prev' = map (\(p1, p2) -> (p1, p2, p2)) prev
+  in moreRestrictivePairAux solver ns prev' (s1, s2)
+
+-- TODO tick adjusting here?
+isIdentity :: (Id, Expr) -> Bool
+isIdentity (i1, Tick _ e2) = isIdentity (i1, e2)
+isIdentity (i1, (Var i2)) = i1 == i2
+isIdentity _ = False
+
+-- approximation should be the identity map
+-- needs to be enforced, won't just happen naturally
+moreRestrictiveEquiv :: S.Solver solver =>
+                        solver ->
+                        HS.HashSet Name ->
+                        StateET ->
+                        StateET ->
+                        W.WriterT [Marker] IO (Maybe (PrevMatch EquivTracker))
+moreRestrictiveEquiv solver ns s1 s2 = do
+  let h1 = expr_env s1
+      h2 = expr_env s2
+      s1' = s1 { expr_env = E.union h1 h2 }
+      s2' = s2 { expr_env = E.union h2 h1 }
+  pm_maybe <- moreRestrictivePair solver ns [(s2', s1')] (s1, s2)
+  case pm_maybe of
+    Nothing -> return Nothing
+    Just (PrevMatch _ _ (hm, _) _) ->
+      if foldr (&&) True (map isIdentity $ HM.toList hm)
+      then return pm_maybe
+      else return Nothing
+
+equivFoldL :: S.Solver solver =>
```

Comment:
Could this be written using foldM?  (Actually, same question applies to a lot of the other fold functions i.e. for induction)